app/scripts/modules/core/src/presentation/forms/FormField.tsx

```diff
@@ -30,106 +28,75 @@ export type IFormFieldProps = IFormFieldValidationProps &
   IFieldLayoutPropsWithoutInput &
   IValidationProps;
 
-interface IFormFieldState {
-  validationMessage: IValidationProps['validationMessage'];
-  validationStatus: IValidationProps['validationStatus'];
-  internalValidators: IValidator[];
+function firstDefined<T>(...values: T[]): T {
+  return values.find(val => !isNil(val));
 }
 
-const ifString = (val: any): string => (isString(val) ? val : undefined);
-const firstDefinedNode = (...values: React.ReactNode[]): React.ReactNode => values.find(val => !isNil(val));
+const { useState, useCallback, useContext, useMemo } = React;
 
-export class FormField extends React.Component<IFormFieldProps, IFormFieldState> implements IFormFieldApi {
-  public static defaultProps: Partial<IFormFieldProps> = {
-    validate: noop,
-    onBlur: noop,
-    onChange: noop,
-    name: null,
-  };
+export function FormField(props: IFormFieldProps) {
+  const { input, layout } = props; // ICommonFormFieldProps
+  const { label, help, required, actions } = props; // IFieldLayoutPropsWithoutInput
+  const { validationMessage: messageProp, validationStatus: statusProp, touched: touchedProp } = props;
+  const { value } = props;
 
-  public state: IFormFieldState = {
-    validationMessage: undefined,
-    validationStatus: undefined,
-    internalValidators: [],
-  };
+  const fieldLayoutPropsWithoutInput: IFieldLayoutPropsWithoutInput = { label, help, required, actions };
 
-  private destroy$ = new Subject();
-  private value$ = new Subject();
+  // Internal validators are defined by an Input component
+  const [internalValidators, setInternalValidators] = useState([]);
+  const addValidator = useCallback((v: IValidator) => setInternalValidators(list => list.concat(v)), []);
+  const removeValidator = useCallback((v: IValidator) => setInternalValidators(list => list.filter(x => x !== v)), []);
 
-  public name = () => this.props.name;
+  const fieldLayoutFromContext = useContext(LayoutContext);
```

Comment:
I'm curious what other Deck state you feel might be good candidate for storing on React context?

Reply:
I'm wary of putting too much stuff in the context because it can make the answer to "where does this data come from?" harder to answer.  There is a similar smell in AngularJS land -- you can put stuff higher up in the `$scope` tree and reference it down below, but tends to be an anti-pattern.  It introduces undocumented, context sensitive dependencies.

Where I do think it makes sense is when you want to scope some setting to a portion of the DOM.  For example, there may be multiple forms on the screen at the same time (each with different Layouts). Prop-drilling (things like layouts) all the way down to the leaf components would be explicit and easy to trace the data flow, but it would also be tedious and introduce excessive boilerplate to the API.

Reply:
Makes sense, thanks as always for the thoughtful response! :heart_eyes_cat: